### PR TITLE
Date helper datetime support

### LINKED ISSUE #69: Support DateTime for all Modules
Support to use `DateTime` as well as `NaiveDateTime`.

Needed to enable option 4 of https://github.com/quantum-elixir/quantum-core/issues/407

### PR DESCRIPTION
**Pull request by shaolang**
@maennchen I've simplified `date_helper` with an extracted `add/3` function. As focus switches to testing that new function rigorously, I have to make it public during tests, I used PragDave's [private](https://hexdocs.pm/private/Private.html) library to keep `add/3` private in non-test environments,

Functions in `date_helper` are now all able to do datetime arithmetic correctly, even for time gaps when daylight savings starts, and ambiguous times when daylight savings ends.

One thing to note is that for ambiguous times when daylight saving ends, I've chosen to return the time with the standard timezone, e.g., between Eastern Daylight Time (EDT) and Eastern Standard Time (EST), `add/3` returns the time with EST. If you think it should return EDT instead, let me know, and I'll change it to that.

Resolves #69

### CODE REVIEW

**Comment by maennchen** [comment]
Thanks a lot for the PR!

This is just the feedback for a quick glance over the code. I'll do a more thorough review later.

* I would like to avoid adding an extra dependency for private functions. Specifying @moduledoc / @doc false should ve sufficient.
* Since the DateHelper module is not part of the public API, we should also not define types there. (Doc link will not work)
* I think triggering twice for ambigous datetimes would be technically correct. But it could make sense making the behavior configurable.

A cron expression, which runs every 5 minutes should also do so when DST changes.

A cron expression running at 2:30 AM every day however should not.

I'm thinking if it would therefore make sense to introduce an option argument where you could choose both, first or last. (Naming tbd)

**Comment by shaolang** [comment]
I've removed private as a dependency and reinstated the date type in `DateChecker`.

Hmmm... ambiguous times are really a headache. I chose to return one of the instances is due to [AWS EventBridge Scheduler's behavior](https://docs.aws.amazon.com/scheduler/latest/UserGuide/schedule-types.html#daylist-savings-time). AWS' scheduer skips running when there's a gap, and runs only once when there's ambiguity.

Nevertheless, you let me know how you'll like to proceed from here. We are getting closer to closing #69 and improving all crontab dependents :)

File: lib/crontab/date_helper.ex [hunk 1]
```diff
@@ -202,33 +158,7 @@ defmodule Crontab.DateHelper do
...
-
-    if leap_year? do
-      # FIXME: How to correct date with tz?
-      DateTime.add(date, -366, :day)
-    else
-      # FIXME: How to correct date with tz?
-      DateTime.add(date, -365, :day)
-    end
-  end
+  def dec_year(date), do: add(date, -days_in_year(date), :day)
```
Comment (maennchen) [inline lib/crontab/date_helper.ex:161]: We might have a bug here. (Also in the implementation before.)

Given date `2024-02-01` (before leap day), I think we subtract one day too much.

Same with increment if we're after the leap day.

File: lib/crontab/date_helper.ex [hunk 2]
```diff
@@ -243,23 +173,12 @@ defmodule Crontab.DateHelper do
 
   """
   @spec inc_month(date) :: date when date: date
-  def inc_month(date = %NaiveDateTime{day: day}) do
+  def inc_month(date = %{year: year, month: month, day: day}) do
     days =
-      date
-      |> NaiveDateTime.to_date()
+      Date.new!(year, month, day)
```
Comment (maennchen) [inline lib/crontab/date_helper.ex:178]:
```suggestion
      date
```

**Review by maennchen — COMMENTED** [review]
> Hmmm... ambiguous times are really a headache. I chose to return one of the instances is due to AWS EventBridge Scheduler's behavior. AWS' scheduer skips running when there's a gap, and runs only once when there's ambiguity.

I think we should let the user decide.

What do you think about adding `on_ambiguous_time_trigger` to the `CronExpression` struct? The values would be `both`, `earlier`, `later` and default to `later`.

Like this the user can decide how he wants to handle this edge-case.

We would then pass it through as an option in a keyword list to the `DateHelper`.

File: lib/crontab/date_helper.ex [hunk 3]
```diff
@@ -362,69 +258,60 @@ defmodule Crontab.DateHelper do
...
+
+  @doc false
+  def add(datetime = %NaiveDateTime{}, amt, unit), do: NaiveDateTime.add(datetime, amt, unit)
+
+  def add(datetime = %DateTime{}, amt, unit) do
+    candidate = DateTime.add(datetime, amt, unit)
+    adjustment = datetime.std_offset - candidate.std_offset
+    adjusted = DateTime.add(candidate, adjustment, :second)
+
+    case adjusted.std_offset == candidate.std_offset do
```
Comment (maennchen) [inline lib/crontab/date_helper.ex:306]: use an if for booleans

File: test/crontab/date_helper_test.exs [hunk 4]
```diff
@@ -4,10 +4,74 @@ defmodule Crontab.DateHelperTest do
...
+        expected =
+          DateTime.new!(~D[2024-11-03], two_plus, "America/New_York")
+          |> DateTime.add(-1, :hour)
+
+        assert DateHelper.add(one_sec_before, 1, unquote(unit)) == expected
+      end
+    end
+
+    setup do
+      Calendar.put_time_zone_database(Tz.TimeZoneDatabase)
```
Comment (maennchen) [inline test/crontab/date_helper_test.exs:74]: We could also put this into a config/config.exs file since that is not read for libraries.

File: lib/crontab/date_helper.ex [hunk 5]
```diff
@@ -118,32 +116,16 @@ defmodule Crontab.DateHelper do
...
-      weekday == 7 -> NaiveDateTime.add(date, -2, :day).day
-      weekday == 6 && previous_day.month == date.month -> previous_day.day
-      weekday == 6 -> NaiveDateTime.add(date, 2, :day).day
-      true -> date.day
-    end
-  end
-
-  def next_weekday_to(date = %DateTime{year: year, month: month, day: day}) do
+  def next_weekday_to(date = %{year: year, month: month, day: day}) do
     weekday = :calendar.day_of_the_week(year, month, day)
```
[the remark this answers is not in the document]
Comment (shaolang) [reply lib/crontab/date_helper.ex:120]: That was in the original code. I'll change it to `Date.day_of_week/1`.

**Comment by shaolang** [comment]
> > Hmmm... ambiguous times are really a headache. I chose to return one of the instances is due to AWS EventBridge Scheduler's behavior. AWS' scheduer skips running when there's a gap, and runs only once when there's ambiguity.
> 
> I think we should let the user decide.
> 
> What do you think about adding `on_ambiguous_time_trigger` to the `CronExpression` struct? The values would be `both`, `earlier`, `later` and default to `later`.
> 
> Like this the user can decide how he wants to handle this edge-case.
> 
> We would then pass it through as an option in a keyword list to the `DateHelper`.

Sounds good to me.

Question then is: does that mean most functions in `DateHelper` need to return either one or two dates?

**Comment by maennchen** [comment]
> does that mean most functions in DateHelper need to return either one or two dates?

Probably not:

If you have the setting on both, we can just return the first occurrence. When we look for the next candidate date, we should automatically reach the later occurrence.

So the logic could be something like this:

* if result ambiguous and setting is `earlier`: return earlier
* if result ambiguous and setting is `later`: return later
* if result ambiguous and setting is `both`: return earlier or later depending on `adjusted`

**Comment by shaolang** [comment]
I've addressed the issues in the review. I've also added targeted tests for `inc_year/1` and `dec_year/1` to ensure the edge cases are covered.

**Comment by maennchen** [comment]
@shaolang Do I understand correct that you want to send a secod PR for the ambiguity habdling?

**Comment by shaolang** [comment]
> @shaolang Do I understand correct that you want to send a secod PR for the ambiguity habdling?

Yes. Are you closing this PR prior to that? I'm okay with keep this or submitting a new PR.

File: lib/crontab/date_helper.ex [hunk 6]
```diff
@@ -278,33 +213,13 @@ defmodule Crontab.DateHelper do
...
-    DateTime.add(date, -(day + max(days_in_last_month - day, 0)), :day)
-  end
-
-  defp day_in_last_month(start_date), do: day_in_last_month(start_date, start_date)
-
-  defp day_in_last_month(date = %Date{month: month}, start_date = %Date{month: month}),
-    do: date |> Date.add(-1) |> day_in_last_month(start_date)
-
-  defp day_in_last_month(date, _start_date), do: date
+  # defp day_in_last_month(%{year: year, month: month}),
```
Comment (maennchen) [inline lib/crontab/date_helper.ex:221]: Can we remove commeted out code?

File: lib/crontab/date_helper.ex [hunk 7]
```diff
@@ -314,25 +229,14 @@ defmodule Crontab.DateHelper do
   defp _beginning_of(date, []), do: date
 
   @spec _end_of(date, [{unit, {any, any}}]) :: date when date: date
-  defp _end_of(date = %NaiveDateTime{}, [{unit, {_, :end_of_month}} | tail]) do
+  defp _end_of(date = %{year: year, month: month, day: day}, [{unit, {_, :end_of_month}} | tail]) do
     upper =
-      date
-      |> NaiveDateTime.to_date()
+      Date.new!(year, month, day)
```
Comment (maennchen) [inline lib/crontab/date_helper.ex:234]:
```suggestion
      date
```

**Review by maennchen — CHANGES_REQUESTED** [review]

File: lib/crontab/date_helper.ex (hunk 6 above)
Comment (shaolang) [reply lib/crontab/date_helper.ex:221]: sorry, my bad.

**Comment by maennchen** [comment]
Separate work for me. But I‘ll merge directly onto `main`. We‘ll just do PR 2 before I cut a release.

Amazing work, thanks a lot for your efforts!

**Review by maennchen — APPROVED** [review]

**Comment by maennchen** [comment]
@shaolang Are you still planning to open a PR for the ambiguous date handling?

I don't really have the time to address it myself and I won't be able to release your work if it is incomplete.

**Comment by shaolang** [comment]
Yes, I do. I have been pretty busy with work since the contribution. I foresee some down time towards the end of the year, so I'll pick it up again and hopefully complete it before the year ends.
